Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/ProgressIndicators/ProgressIndicatorLoadingBar.cs

```diff
@@ -0,0 +1,97 @@
+
+using System.Threading.Tasks;
+using TMPro;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.SceneTransitions
+{
+    /// <summary>
+    /// Displays progress bar and (optionally) percentage text.
+    /// </summary>
+    [ExecuteAlways]
+    public class ProgressIndicatorLoadingBar : MonoBehaviour, IProgressIndicator
+    {
+        const float SmoothProgressSpeed = 0.25f;
+
+        public Transform MainTransform { get { return transform; } }
+        public ProgressIndicatorState State { get { return state; } }
+        public float Progress { set { targetProgress = value; } }
+        public string Message { set { messageText.text = value; } }
+
+        // The animated progress bar object
+        [SerializeField]
+        private Transform progressBar = null;
+
+        // The message text used by the 'Visible' message style
+        [SerializeField]
+        private TextMeshPro messageText = null;
+
+        // The progress text used by all non-'None' progress styles
+        [SerializeField]
+        private TextMeshPro progressText = null;
+
+        [Tooltip("Format to use when converting progress number to string.")]
+        [SerializeField]
+        private string progressStringFormat = "{0:P0}";
+
+        [Tooltip("Whether to display the percentage as text in addition to the loading bar.")]
+        [SerializeField]
+        private bool displayPercentage = true;
+        
+        private float smoothProgress = 0f;
+        private ProgressIndicatorState state = ProgressIndicatorState.Closed;
+        private Vector3 barScale = Vector3.one;
+
+        /// <summary>
+        /// Target for smooth progress, 0-1. Serialized to test in editor.
+        /// </summary>
+        [SerializeField]
+        [Range(0f, 1f)]
+        private float targetProgress = 0f;
+        
+        public async Task OpenAsync()
+        {
+            gameObject.SetActive(true);
+
+            state = ProgressIndicatorState.Opening;
+
+            await Task.Yield();
+
+            state = ProgressIndicatorState.Open;
+        }
+
+        public async Task CloseAsync()
+        {
+            state = ProgressIndicatorState.Closing;
+
+            await Task.Yield();
+
+            state = ProgressIndicatorState.Closed;
+
+            gameObject.SetActive(false);
+        }
+
+        private void Update()
+        {
+            if (state != ProgressIndicatorState.Open)
+                return;
+
+            smoothProgress = Mathf.Lerp(smoothProgress, targetProgress, SmoothProgressSpeed);
+
+            if (smoothProgress > 0.99f)
+                smoothProgress = 1f;
+
+            barScale.x = Mathf.Clamp(smoothProgress, 0.01f, 1f);
+            progressBar.localScale = barScale;
+
+            progressText.gameObject.SetActive(displayPercentage);
+            progressText.text = string.Format(progressStringFormat, smoothProgress);
```

Comment:
I think that this might allocate a new string on every single Update() - it might be worth caching the last value IF it's still the same, or otherwise doing some work to pre-allocate the strings that could render.

@Cameron-Micka had similar things with the visual profiler.

Reply:
FWIW, pre-allocate could be overkill, at least caching would be great.